### DISCUSSION TITLE
HSTS Middleware

### DISCUSSION THREAD

**Discussion by yefetBenTili** [fastapi/fastapi#4092] [Questions]
Hi community.

For security purposes, did anyone had to enforce [HSTS strict-transport-security](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Strict-Transport-Security) headers in all responses in their app?

I am not sure if Fastapi support HSTS by default or if there is a Middleware in order to send all the HSTS headers to clients?

Or maybe this could  be a new middleware project

**Comment by acidjunk** [comment]
I wonder if this kind of setup can' be done by a reverse proxy like NginX or Traeffik: https://stackoverflow.com/questions/51300371/how-to-configure-hsts-headers-in-traefik

In most deployments I do https off loading before a request reaches the app. I am not sure if this handles/covers all the HSTS details; but if you can enforce this from a webserver in front of FastAPI it will make the resulting app simpeler (and simpler to run locally without HSTS)

**Comment by jm-nab** [comment]
You can do something like this:

```
@app.middleware("http")
async def add_security_headers(request, call_next):
        response = await call_next(request)
        response.headers[
            "Strict-Transport-Security"
        ] = "max-age=63072000; includeSubDomains"
        # TODO enable CSP cause this will probably break whatever you currently have
        # response.headers["Content-Security-Policy"] = "default-src 'self'; script-src 'self'; style-src 'self' https: 'unsafe-inline';"
        response.headers["X-Frame-Options"] = "DENY"
        response.headers["X-Content-Type-Options"] = "nosniff"
        return response
```
